tempesta_fw/sock_srv.c

```diff
@@ -539,83 +550,404 @@ tfw_srv_conn_free(TfwSrvConn *srv_conn)
+	TfwServer *srv = psrv;
 
 	list_for_each_entry_safe(srv_conn, tmp, &srv->conn_list, list) {
 		tfw_connection_unlink_from_peer((TfwConn *)srv_conn);
 		tfw_srv_conn_free(srv_conn);
 	}
+}
+
+static int
+tfw_sock_srv_start_srv(TfwServer *srv)
+{
+	int r;
+
+	TFW_DBG_ADDR("start server", &srv->addr);
+
+	if ((r = tfw_sock_srv_add_conns(srv)))
+		return r;
+	if ((r = tfw_apm_add_srv(srv)))
+		return r;
+	if ((r = tfw_sock_srv_connect_srv(srv)))
+		return r;
+
+	return 0;
+}
+
+/*
+ * ------------------------------------------------------------------------
+ *	Grace shutdown operations
+ * ------------------------------------------------------------------------
+ */
+/* Grace shutdown timeout. */
+static unsigned int tfw_cfg_grace_time = 0;
+/* List of server on grace shutdown. */
+static LIST_HEAD(tfw_gs_servers);
+static DEFINE_SPINLOCK(tfw_gs_lock);
+
+static void
+tfw_sock_srv_grace_list_add(TfwServer *srv)
+{
+	spin_lock(&tfw_gs_lock);
+
+	tfw_server_get(srv);
+	list_add(&srv->list, &tfw_gs_servers);
+
+	spin_unlock(&tfw_gs_lock);
+}
+
+static void
+tfw_sock_srv_grace_list_del(TfwServer *srv)
+{
+	spin_lock(&tfw_gs_lock);
+
+	list_del_init(&srv->list);
+	tfw_server_put(srv);
+
+	spin_unlock(&tfw_gs_lock);
+}
+
+static void
+tfw_sock_srv_grace_stop(TfwServer *srv)
+{
+	tfw_server_stop_sched(srv);
+	tfw_sock_srv_disconnect_srv(srv);
+	tfw_sock_srv_grace_list_del(srv);
+}
+
+static void
+tfw_sock_srv_grace_shutdown_cb(unsigned long data)
+{
+	TfwServer *srv = (TfwServer *)data;
+
+	tfw_sock_srv_grace_stop(srv);
+}
+
+/**
+ * Schedule graceful shutdown of a server. Allow server to finish it's
+ * forward queue or pinned sessions if any.
+ *
+ * The function is called under server group lock. @sg->srv_list is changed
+ * during this function.
+ */
+static int
+tfw_sock_srv_grace_shutdown_srv(TfwSrvGroup *sg, TfwServer *srv)
+{
+	int r = 0;
+
+	tfw_server_get(srv);
+	__tfw_sg_del_srv(sg, srv, false);
+	srv->flags |= TFW_CFG_F_DEL;
+
+	if (!tfw_cfg_grace_time) {
+		r = tfw_sock_srv_disconnect_srv(srv);
+	} else {
+		if ((srv->sg->flags & TFW_SRV_STICKY) &&
+		    atomic64_read(&srv->sess_n))
+		{
+			tfw_server_start_sched(srv);
+		}
+		setup_timer(&srv->gs_timer, tfw_sock_srv_grace_shutdown_cb,
+			    (unsigned long)srv);
+		tfw_sock_srv_grace_list_add(srv);
+		mod_timer(&srv->gs_timer, jiffies + tfw_cfg_grace_time * HZ);
+	}
+	tfw_server_put(srv);
+
+	return r;
+}
+
+static int
+tfw_sock_srv_grace_shutdown_sg(TfwSrvGroup *sg)
+{
+	TfwServer *srv, *srv_tmp;
+
+	tfw_sg_stop_sched(sg);
+
+	write_lock(&sg->lock);
+	list_for_each_entry_safe(srv, srv_tmp, &sg->srv_list, list)
+		tfw_sock_srv_grace_shutdown_srv(sg, srv);
+	write_unlock(&sg->lock);
+
+	tfw_sg_put(sg);
 
 	return 0;
 }
 
+/**
+ * Stop grace shutdown timers and stop servers.
+ */
 static void
-tfw_sock_srv_delete_all_conns(void)
+tfw_sock_srv_grace_shutdown_now(void)
 {
-	tfw_sg_for_each_srv(tfw_sock_srv_del_conns);
+	while (1)
+	{
+		TfwServer *srv;
+		int act;
```

Comment:
Please place open curly brace on a next line only in case of complex multi-line conditions and place it on the same line in all other cases.

Do you really need `act` variable? Doesn't
```C           
                if (del_timer_sync(&srv->gs_timer))                                   
                        tfw_sock_srv_grace_stop(srv);
```
look simpler?